### ISSUE TITLE
SC1036 for function declaration "==== () { echo test; }"

### ISSUE THREAD

**Issue by kbauer** [koalaman/shellcheck#2047]
## How to reproduce

Give ShellCheck the following script:

    #!/usr/bin/env bash
    ==== () { echo "HEADING: $1"; }

## Seen output

```
Line 2:
==== () { echo "HEADING: $1"; }
     ^-- SC1036: '(' is invalid here. Did you forget to escape it?
     ^-- SC1088: Parsing stopped here. Invalid use of parentheses?
```

## Expected output

None. In Bash scripts, `====` is a valid identifier. This can be used e.g. to combine structuring a script with headings and producing progress output at the same time.

However, it looks like this is very specific to Bash. Both `ksh` and `sh` complain about different syntax errors caused by an invalid identifier.

**Comment by matthewpersico** [comment]
This is a bash bug. This should be illegal: https://stackoverflow.com/questions/28114999/what-are-the-rules-for-valid-identifiers-e-g-functions-vars-etc-in-bash .

**Comment by kbauer** [comment]
It's really not entirely clear; The discussion also shows that this interpretation of the documentation doesn't match the actual behavior of bash, with backwards-compatibility requirements making future changes unlikely. Bash *does* however reject the identifier when invoked as `bash --posix`. So maybe it should be allowed, but warned about?

On the other hand, `ShellCheck` doesn't catch the use of an invalid identifier in:
```
#!/usr/bin/env ksh
hello-world () { :; }
```

**Comment by koalaman** [comment]
As a workaround, you can use `function ==== () { echo "HEADING: $1"; }`.

The general problem has been that bash accepts a lot of strange names, e.g. `**=??!`, and ShellCheck tends to give misleading advice when any such strings could be considered the start of a function.

To avoid suggesting "Maybe you're missing () in a function" whenever you write random gibberish, it therefore only allows particularly weird characters after a `function` keyword even though bash doesn't care as much.

**Comment by matthewpersico** [comment]
> On the other hand, `ShellCheck` doesn't catch the use of an invalid identifier in:
> 
> ```
> #!/usr/bin/env ksh
> hello-world () { :; }
> ```

Assuming you are talking about the function name, shellcheck allows hello-world probably because, although invalid according time the docs, it is works in bash to have dashes in function names. Not identifiers though. If you’re talking about the :;, then disregard.

**Comment by koalaman** [comment]
It does warn for sh and dash as part of the bashism checks. It's simply missing the same for ksh.

**Comment by navid-zamani** [comment]
I got the same error on my definition of a function:

```bash
~~() { cd "$importantDirectory";  }
```

And I disagree that it is a ”bug”. It is a feature. And was done like this deliberately too.

What’s a bug, in any language, is when identifiers are artificially limited beyond actually reserved symbols and symbol categories. (Worse, bash completely fails at handling identifiers with any characters outside of English/Latin. … The world is not English. For the vast majority of people, Unicode is a fundamental necessity. But, having reported this, the maintainers of bash are… weird, to say the least. One day, I might write a patch myself…)
